user/session.go

```diff
@@ -55,13 +55,41 @@ type APILimit struct {
 	Smoothing *apidef.RateLimitSmoothing `json:"smoothing" bson:"smoothing"`
 }
 
+// Clone does a deepcopy of APILimit.
+func (a APILimit) Clone() APILimit {
+	clone := APILimit{
+		Rate:               a.Rate,
+		Per:                a.Per,
+		ThrottleInterval:   a.ThrottleInterval,
+		ThrottleRetryLimit: a.ThrottleRetryLimit,
+		MaxQueryDepth:      a.MaxQueryDepth,
+		QuotaMax:           a.QuotaMax,
+		QuotaRenews:        a.QuotaRenews,
+		QuotaRemaining:     a.QuotaRemaining,
+		QuotaRenewalRate:   a.QuotaRenewalRate,
+		SetBy:              a.SetBy,
+	}
+
+	if a.Smoothing != nil {
+		clone.Smoothing = &apidef.RateLimitSmoothing{
```

Comment:
A shallow copy can be used here, something like:
```
copied := *a.Smoothing
clone.Smoothing = &copied
```
Alternatively, implement Clone on the type as well.